### PR TITLE
Fixes in order to get the local indexing to work

### PR DESCRIPTION
**Pull request by mostaphaRoudsari**
There are two commits that are related to development issues here:

- One that fixes the path issues for Windows path
- A second one that fixes the path for accessing the local resources

With these fixes I could create a local index and get `queenbee repo list` to work. Once this is merged I need to:

1. Send a new PR to honeybee-radiance-recipe repository with the new file structure.
2. Add a new command to Queenbee luigi to support the new way of getting the recipe from a repository instead of a baked recipe file.
3. Use the new command the Grasshopper component to deal with honeybee-radiance recipes.

### CODE REVIEW

File: queenbee/base/request.py [hunk 1]
```diff
@@ -1,19 +1,34 @@
...
 
 
+def get_uri(url):
+    """Resolve uri for urls and local files."""
+
+    if url.startswith('file:///'):
+        return url.replace('\\', '/')
+
+    try:
+        uri = pathlib.Path(os.path.abspath(url)).as_uri()
```
Comment (AntoineDao) [inline queenbee/base/request.py:17]: Why are we trying to resolve an http url as an absolute path? Lines 13-14 address the case where a url is for a local path so maybe we should make that one an `abspath`.

What happens if we keep this line is that all the `http` urls are appended with a local file path like so:

```
file:///root/projects/ladybug-tools/queenbee/https%3A/api.pollination.cloud/registries/antoinedao
```

File: queenbee/base/request.py [hunk 2]
```diff
@@ -1,19 +1,34 @@
...
+
 def urljoin(*args):
     """join multiple string into a clean url
 
     Returns:
         str: a clean url string
     """
     url = "/".join(map(lambda x: str(x).rstrip('/'), args))
-    url.replace('\\', '/')
+    url = url.replace('\\', '/')
```
Comment (AntoineDao) [inline queenbee/base/request.py:31]: Good catch... 😅

File: queenbee/cli/repository/init.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def init(path):
 
     path = os.path.abspath(path)
 
-    os.makedirs(os.path.join(path, 'workflows'), exist_ok=True)
+    os.makedirs(os.path.join(path, 'operators'), exist_ok=True)
```
Comment (AntoineDao) [inline queenbee/cli/repository/init.py:23]: Good catch... 😅

File: queenbee/cli/repository/manage.py [hunk 4]
```diff
@@ -26,12 +25,6 @@ def add(name, path, force):
     """
     ctx = click.get_current_context()
 
-    parse_res = urlparse(path)
```
Comment (AntoineDao) [inline queenbee/cli/repository/manage.py:29]: Thanks for dedupelicating this 😀

File: queenbee/config/repositories.py [hunk 5]
```diff
@@ -37,9 +31,7 @@ def fetch(self, auth_header: str = '') -> 'RepositoryIndex':
         from ..repository import RepositoryIndex
 
         if self.path.startswith('file:'):
-            path = self.path.split('file:')[-1]
-            index_path = os.path.join(path, 'index.json')
-            url = f'file:{index_path}'
+            url = os.path.join(self.path, 'index.json').replace('\\', '/')
```
Comment (AntoineDao) [inline queenbee/config/repositories.py:34]: If we are already sanitising the path input when validating the `RepositoryReference.path` value do we still need to use `replace('\\', '/')`?

File: queenbee/repository/package.py [hunk 6]
```diff
@@ -274,12 +274,12 @@ def from_package(cls, package_path: str):
 
     def fetch_package(self, source_url: str = None, verify_digest: bool = True, auth_header: str = '') -> 'PackageVersion':
         if source_url.startswith('file:'):
-            source_path = source_url.split('file:')[1]
-
+            source_path = source_url.split('file:///')[1]
+            subfolder = f'{self.type}s'
             if os.path.isabs(source_path):
-                package_path = os.path.join(source_path, self.url)
+                package_path = os.path.join(source_path, subfolder, self.url)
```
Comment (AntoineDao) [inline queenbee/repository/package.py:280]: We don't need to add the `subfolder` here. The `self.url` is be a valid relative path from the `source_path`.

image

**Review by AntoineDao — CHANGES_REQUESTED** [review]
Thanks for the catches and for refactoring the path management functions. Two things are breaking when I run the code locally so we should change them or discuss them for me to better understand what you're after:

1. Trying to create an absolute path from an `http` url in `queenbee/base/request.py` creates a double path which is invalid
2. There should be no need to add a `subfolder` value to a `package_path` when fetching from a repository because the package `url` should be a valid relative path from the repository `index.json`

File: queenbee/base/request.py (hunk 1 above)
Comment (mostaphaRoudsari) [reply queenbee/base/request.py:17]: Good catch! I will fix this.

File: queenbee/config/repositories.py (hunk 5 above)
Comment (mostaphaRoudsari) [reply queenbee/config/repositories.py:34]: You are right. This one is redundant knowing the the path is already cleaned. Let me remove this one.

File: queenbee/repository/package.py (hunk 6 above)
Comment (mostaphaRoudsari) [reply queenbee/repository/package.py:280]: This is very different from what I get for a local repository:

image

We need the subfolder to get the local ones to work.

Comment (mostaphaRoudsari) [reply queenbee/repository/package.py]: This commit should fix the bug for adding subfolder name to the url.

https://github.com/ladybug-tools/queenbee/pull/148/commits/e4928355a29aafa2832f8b0b6b961dac0d52e083

**Comment by mostaphaRoudsari** [comment]
@AntoineDao, both of your comments should be addressed. Let me know if you have any other comments.

File: queenbee/base/request.py (hunk 1 above)
Comment (AntoineDao) [reply queenbee/base/request.py:20]: Still not sure this is resolved. From my understanding a `url` should only start with either: `file://`, `http://` or `https://`. Therefore lines 18/24 just look like a no-op to me.

**Review by AntoineDao — APPROVED** [review]
Still not 100% sure what's happening in `queenbee/base/request.py` as per the comment I unresolved. Other than that it's looking good! 👍

File: queenbee/base/request.py (hunk 1 above)
Comment (mostaphaRoudsari) [reply queenbee/base/request.py:20]: You hit line 18-24 the first time that you add a local repository. In that case `url` is a file but doesn't start with a `file:///`.

`queenbee repo add honeybee-radiance usr/path/to/local/repo`

Comment (AntoineDao) [reply queenbee/base/request.py:20]: Oh! Great catch! 😅

**Pull request merged**